File: MiniDB_Projects/Team_QueryCrafters/src/storage/buffer_pool.py
from typing import Dict, List, Tuple
from src.storage.page import Page

class BufferPool:
    def __init__(self, capacity: int = 10, wal_manager=None):
        self.capacity = capacity
        self.wal_manager = wal_manager
        # Mapping from (table_name, page_id) -> Dict containing page, pin_count, is_dirty, heap_file
        self.pool = {}
        # LRU order: list of (table_name, page_id) keys
        self.lru_order = []

    def fetch_page(self, page_id: int, heap_file) -> Page:
        key = (heap_file.table_name, page_id)
        
        if key in self.pool:
            entry = self.pool[key]
            entry["pin_count"] += 1
            # Move to end of LRU list (most recently used)
            self.lru_order.remove(key)
            self.lru_order.append(key)
            return entry["page"]

        # If not in pool, check if capacity is reached
        if len(self.pool) >= self.capacity:
            # Find an eviction candidate (pin_count == 0) using LRU order
            evict_key = None
            for k in self.lru_order:
                if self.pool[k]["pin_count"] == 0:
                    evict_key = k
                    break
            
            if evict_key is None:
                raise RuntimeError("Buffer pool is full and all pages are pinned (cannot evict)")

            # Evict candidate
            self.evict_page(evict_key)

        # Read page from disk
        page_data = heap_file.read_page_from_disk(page_id)
        page = Page(page_id)
        page.deserialize(page_data)

        self.pool[key] = {
            "page": page,
            "pin_count": 1,
            "is_dirty": False,
            "heap_file": heap_file
        }
        self.lru_order.append(key)
        return page

    def unpin_page(self, page_id: int, heap_file, is_dirty: bool = False):
        key = (heap_file.table_name, page_id)
        if key in self.pool:
            entry = self.pool[key]
            entry["pin_count"] = max(0, entry["pin_count"] - 1)
            if is_dirty:
                entry["is_dirty"] = True
            
            # Update LRU order: move to end
            if key in self.lru_order:
                self.lru_order.remove(key)
            self.lru_order.append(key)

    def mark_dirty(self, page_id: int, heap_file):
        key = (heap_file.table_name, page_id)
        if key in self.pool:
            self.pool[key]["is_dirty"] = True

    def evict_page(self, key: Tuple[str, int]):
        entry = self.pool[key]
        if entry["is_dirty"]:
            self.flush_entry(key, entry)
        
        del self.pool[key]
        self.lru_order.remove(key)

    def flush_entry(self, key: Tuple[str, int], entry: dict):
        # Write-Ahead Logging rule: flush WAL to disk before writing the dirty page
        if self.wal_manager is not None:
            self.wal_manager.flush()
        
        table_name, page_id = key
        entry["heap_file"].write_page_to_disk(page_id, entry["page"].serialize())
        entry["is_dirty"] = False

    def flush_page(self, page_id: int, heap_file):
        key = (heap_file.table_name, page_id)
        if key in self.pool:
            entry = self.pool[key]
            if entry["is_dirty"]:
                self.flush_entry(key, entry)

    def flush_all(self):
        for key, entry in list(self.pool.items()):
            if entry["is_dirty"]:
                self.flush_entry(key, entry)

File: MiniDB_Projects/Team_minions/src/storage/buffer_pool.cpp
#include "minidb/storage/buffer_pool.h"

#include "minidb/exceptions.h"

namespace minidb {

BufferPool::BufferPool(std::size_t capacity) : capacity_(capacity) {
    if (capacity_ == 0) capacity_ = 1;
}

int BufferPool::register_file(DiskManager* dm) {
    files_.push_back(dm);
    return static_cast<int>(files_.size()) - 1;
}

void BufferPool::register_file_with_id(int file_id, DiskManager* dm) {
    if (static_cast<int>(files_.size()) <= file_id) {
        files_.resize(file_id + 1, nullptr);
    }
    files_[file_id] = dm;
}

page_id_t BufferPool::file_page_count(int file_id) const {
    return files_.at(file_id)->num_pages();
}

Page* BufferPool::fetch_page(int file_id, page_id_t page_id) {
    uint64_t key = make_key(file_id, page_id);
    auto it = page_table_.find(key);
    if (it != page_table_.end()) {
        // Cache hit. Pin it and remove it from the LRU list (pinned frames are
        // not eviction candidates).
        Frame* frame = it->second.get();
        if (frame->pin_count == 0) {
            auto pos = lru_pos_.find(key);
            if (pos != lru_pos_.end()) {
                lru_.erase(pos->second);
                lru_pos_.erase(pos);
            }
        }
        frame->pin_count++;
        ++hits_;
        return &frame->page;
    }

    // Cache miss: make room, then read the page from disk into a new frame.
    ++misses_;
    if (page_table_.size() >= capacity_) {
        evict_one();
    }
    auto frame = std::make_unique<Frame>();
    frame->file_id = file_id;
    frame->page_id = page_id;
    frame->pin_count = 1;
    frame->dirty = false;
    std::vector<uint8_t> raw;
    files_.at(file_id)->read_page(page_id, raw);
    frame->page = Page(raw);
    Page* result = &frame->page;
    page_table_[key] = std::move(frame);
    return result;
}

Page* BufferPool::new_page(int file_id, page_id_t* out_page_id) {
    page_id_t pid = files_.at(file_id)->allocate_page();
    if (out_page_id) *out_page_id = pid;
    // The page was zero-filled on disk; fetch it so it lives in a frame, then
    // overwrite it with a properly initialised (empty slotted) page.
    Page* page = fetch_page(file_id, pid);
    *page = Page();  // reset header to an empty slotted page
    return page;
}

void BufferPool::unpin_page(int file_id, page_id_t page_id, bool is_dirty) {
    uint64_t key = make_key(file_id, page_id);
    auto it = page_table_.find(key);
    if (it == page_table_.end()) return;
    Frame* frame = it->second.get();
    if (is_dirty) frame->dirty = true;
    if (frame->pin_count > 0) frame->pin_count--;
    if (frame->pin_count == 0) {
        // Newly evictable: append to the back of the LRU list (most recent).
        lru_.push_back(key);
        lru_pos_[key] = std::prev(lru_.end());
    }
}

void BufferPool::write_frame_to_disk(Frame& frame) {
    if (!frame.dirty) return;
    // Write-ahead logging: the log must be on disk up to this page's LSN
    // *before* the page itself is written.
    if (flush_log_up_to_) {
        flush_log_up_to_(frame.page.lsn());
    }
    files_.at(frame.file_id)->write_page(frame.page_id, frame.page.bytes());
    frame.dirty = false;
}

void BufferPool::flush_page(int file_id, page_id_t page_id) {
    uint64_t key = make_key(file_id, page_id);
    auto it = page_table_.find(key);
    if (it == page_table_.end()) return;
    write_frame_to_disk(*it->second);
}

void BufferPool::flush_all() {
    for (auto& kv : page_table_) {
        write_frame_to_disk(*kv.second);
    }
}

void BufferPool::evict_one() {
    // Pick the least-recently-used *unpinned* frame (front of the LRU list).
    if (lru_.empty()) {
        throw BufferPoolFullException(
            "buffer pool is full and every page is pinned");
    }
    uint64_t victim_key = lru_.front();
    lru_.pop_front();
    lru_pos_.erase(victim_key);

    auto it = page_table_.find(victim_key);
    if (it == page_table_.end()) return;  // should not happen
    Frame* frame = it->second.get();
    write_frame_to_disk(*frame);  // flush if dirty (respects WAL)
    page_table_.erase(it);
}

}  // namespace minidb
